pkg/controller/sparkapplication/controller.go

```diff
@@ -615,61 +613,57 @@ func (c *Controller) submitSparkApplication(app *v1alpha1.SparkApplication) *v1a
-				State: v1alpha1.SubmittedState,
-			},
-			SubmissionAttempts: app.Status.SubmissionAttempts + 1,
-			ExecutionAttempts:  app.Status.ExecutionAttempts + 1,
-			SubmissionTime:     metav1.Now(),
-		}
-		c.recordSparkApplicationEvent(app)
-
-		// Add driver as a finalizer to prevent SparkApplication deletion till driver is deleted.
-		if app.ObjectMeta.Finalizers == nil {
-			app.ObjectMeta.Finalizers = []string{sparkDriverRole}
 		} else {
-			app.ObjectMeta.Finalizers = append(app.ObjectMeta.Finalizers, sparkDriverRole)
+			glog.Errorf("failed to run spark-submit for SparkApplication %s in namespace: %s. Error: %v", submission.name,
+				submission.namespace, errorMsg)
+			app.Status = v1alpha1.SparkApplicationStatus{
+				AppState: v1alpha1.ApplicationState{
+					State:        v1alpha1.FailedSubmissionState,
+					ErrorMessage: errorMsg,
+				},
+				SubmissionAttempts:        app.Status.SubmissionAttempts + 1,
+				LastSubmissionAttemptTime: metav1.Now(),
+			}
+			c.recordSparkApplicationEvent(app)
+			return app
 		}
+	}
 
-		if app.Spec.Monitoring != nil && app.Spec.Monitoring.Prometheus != nil {
-			// configPrometheusMonitoring may update app.Spec.
-			configPrometheusMonitoring(app, c.kubeClient)
-		}
+	// Submission Successful or Driver Pod Already exists.
+	glog.Infof("spark-submit completed for SparkApplication %s in namespace %s", submission.name, submission.namespace)
+	// Update AppStatus to submitted.
+	app.Status = v1alpha1.SparkApplicationStatus{
+		AppState: v1alpha1.ApplicationState{
+			State: v1alpha1.SubmittedState,
+		},
+		SubmissionAttempts:        app.Status.SubmissionAttempts + 1,
+		ExecutionAttempts:         app.Status.ExecutionAttempts + 1,
+		LastSubmissionAttemptTime: metav1.Now(),
+	}
+	c.recordSparkApplicationEvent(app)
 
-		// Create Spark UI Service.
-		service, err := createSparkUIService(app, c.kubeClient)
-		if err != nil {
-			glog.Errorf("Failed to create UI service for SparkApplication %s/%s. Error: %v", app.Namespace, app.Name, err)
-		} else {
-			app.Status.DriverInfo.WebUIServiceName = service.serviceName
-			app.Status.DriverInfo.WebUIPort = service.nodePort
-			// Create UI Ingress if ingress-format is set.
-			if c.ingressUrlFormat != "" {
-				ingress, err := createSparkUIIngress(app, *service, c.ingressUrlFormat, c.kubeClient)
-				if err != nil {
-					glog.Errorf("Failed to create UI Ingress for SparkApplication %s/%s. Error: %v", app.Namespace, app.Name, err)
-				} else {
-					app.Status.DriverInfo.WebUIIngressAddress = ingress.ingressUrl
-					app.Status.DriverInfo.WebUIIngressName = ingress.ingressName
-				}
+	// Add driver as a finalizer to prevent SparkApplication deletion till driver is deleted.
+	app = addFinalizer(app, sparkDriverRole)
+	if app.Spec.Monitoring != nil && app.Spec.Monitoring.Prometheus != nil {
+		// configPrometheusMonitoring may update app.Spec.
+		configPrometheusMonitoring(app, c.kubeClient)
+	}
+
+	// Create Spark UI Service.
```

Comment:
It's also better to create the UI service before submitting the application to run, which is currently the case in the master.

Reply:
I am not clear about what we gain by creating a service upfront.  The service won't even be up till the driver pod is actually running and start listening on the pod. Also the submission can fail and we will need additionally clean-up if the spark-submit failed but we already created the service/ingress.  Let me know if there are any obvious benefits of doing this and I can change this.

Reply:
While, there's no real benefit on doing this. But if for some reason the service is not up when the driver is up and running, the UI is not accessible. But the chance of that is probably negligible.